readers/combined.cc

```diff
@@ -83,12 +83,17 @@ class mutation_fragment_merger {
     Producer _producer;
     range_tombstone_change_merger<stream_id_t> _tombstone_merger;
     mutation_fragment_v2_opt _result;
+    // An optional histogram describing a distribution of clustering keys.
+    // The map gathers a number of clustering keys merged (value) from a given
+    // number of sstable files (key).
+    std::unordered_map<int32_t, int64_t>* _histogram{ nullptr };
```

Comment:
We have histogram types in `util/histogram.hh`, please see if one of those are a good fit here.

Reply:
Absolutely

Reply:
I think it's a good idea to wrap it in a `struct combined_reader_statistics`, to clarify the intent, and make room for more options. It's probably faster to make it non-optional.

Reply:
Agree. I realised there is a second associated task that requires gathering more information from the readers. Will wrap in an additional struct. 
I made it optional, as this information is only required by a regular compaction. If it is not a problem to collect data with every call to the make_combined_reader, then it can be non-optional as well.

Reply:
> We have histogram types in `util/histogram.hh`, please see if one of those are a good fit here.

Looked at the file and did not find any suitable types for this purpose. Thus I left the map. I think a simple map is a better solution for this purpose. Especially when wrapped in an additional structure.

Reply:
On modern processors additions are free but if ()s are expensive.

The map is a terrible structure for this. Compared to the addition, the lookup and possible allocation are very expensive.

A good type for this is `approx_estimated_histogram<8, 64, 8>`. It will lookup the bucket using just a few instructions.

Reply:
According to the cassandra's nodetool compactionhistory [documentation](https://docs.datastax.com/en/cassandra-oss/3.x/cassandra/tools/toolsCompactionHistory.html), the statistic in the `rows_merged` field has a notation `{tables:rows}`, e.g. `{1:3, 3:1}` means 3 rows were taken from one SSTable (1:3) and 1 row taken from 3 SSTables (3:1). 

If we want to stick to the same format, then a hash map I used in the implementation is sufficient for that. 

On the other hand, if we want a proper histogram and aggregate data into buckets then I agree `approx_exponential_histogram` could be a better choice. That will result in a different output, however. Take the example above. The values 1 and 3 can get into one bucket and then its value would be 4 and we would have something like `{[1:4): 4}`

Reply:
We can make it simple. We know in advance how many sstables we are compacting, so we can allocate a vector of size n+1 and increment the appropriate slot.